test/quantize_cpu.cpp

```diff
@@ -0,0 +1,113 @@
+/*******************************************************************************
+* Copyright 2018 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include <algorithm>
+#include <cinttypes>
+#include <cmath>
+#include <cstdlib>
+#include <string>
+#include "gtest/gtest.h"
+#include "ngraph/ngraph.hpp"
+#include "ngraph/op/constant.hpp"
+#include "ngraph/op/get_output_element.hpp"
+#include "ngraph/runtime/cpu/op/quantize.hpp"
+#include "util/all_close.hpp"
+#include "util/all_close_f.hpp"
+#include "util/ndarray.hpp"
+#include "util/random.hpp"
+#include "util/test_control.hpp"
+#include "util/test_tools.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+TEST(quantize_cpu, quantize_to_uint8_small)
+{
+    vector<float> a_data = {-85.0, 0.0, 2.0, 10.0, 15.0};
+    Shape shape_a{5};
+    auto A = make_shared<op::Parameter>(element::f32, shape_a);
+    auto B = op::Constant::create(element::f32, Shape{}, {-85.0f});
+    auto C = op::Constant::create(element::f32, Shape{}, {15.0f});
+    auto QT = make_shared<op::Quantize>(A, B, C, element::u8);
+    auto output_data = std::make_shared<op::GetOutputElement>(QT, 0);
+    auto output_min = std::make_shared<op::GetOutputElement>(QT, 1);
+    auto output_max = std::make_shared<op::GetOutputElement>(QT, 2);
+    auto f = make_shared<Function>(NodeVector{output_data, output_min, output_max},
+                                   op::ParameterVector{A});
+    auto backend = runtime::Backend::create("CPU");
+    // Create some tensors for input/output
+    auto a = backend->create_tensor(element::f32, shape_a);
+    copy_data(a, a_data);
+    auto result = backend->create_tensor(element::u8, shape_a);
+    auto result_min = backend->create_tensor(element::f32, Shape{});
+    auto result_max = backend->create_tensor(element::f32, Shape{});
+    backend->call(f, {result, result_min, result_max}, {a});
+    EXPECT_EQ((vector<uint8_t>{0, 0, 6, 30, 45}), read_vector<uint8_t>(result));
```

Comment:
Question only: it looks like when quantizing to unsigned, we set the output range to [0,max(abs(input))] but then truncate negative input values to 0 before actually quantizing?

Reply:
Yes, negative values should map to 0